`loginctl` user lingering configuration revisited

tomeon opened the pull request:
## Description of changes

Follow-up to #260248 (and related to #3702).  The main goal is to address the concerns expressed in [this review comment](https://github.com/NixOS/nixpkgs/pull/260248#discussion_r1359637530); namely, that user lingering settings should be left alone when `users.mutableUsers` is enabled.

Questions for reviewers:

1. Error handling in `update-users-groups.pl`.  This script warns, but does not abort, if something goes wrong with enabling or disabling lingering (e.g. failure to create or remove files under `/var/lib/systemd/linger`) (at least, this is true when using the fallback implementations of `do{En,Dis}ableLinger`).  Should these warnings be promoted to fatal exceptions?
2. Does it make sense to have fallback implementation of `do{En,Dis}ableLinger`?  IIUC `/run/booted-system/sw/bin/loginctl` is not guaranteed to exist when the relevant activation script runs.
3. Is `/run/booted-system/sw/bin/loginctl` the correct choice of `loginctl` path?  Should `update-users-groups.pl` instead use (for instance) `${pkgs.systemd}/bin/loginctl`?

Note that I incorporated the user lingering management implementation into `update-users-groups.pl` because this makes it easy to detect whether a given user account is newly-created or not, and thereby determine whether it is okay to touch the user's lingering setting when `users.mutableUsers` is enabled.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`) (kinda sorta non-applicable, though I did test the functionality of the executable [`update-users-groups.pl`])
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tomeon committed:
nixos/users-groups: handle loginctl lingering in update-users-groups.pl

in order to respect `users.mutableUsers = true` by only enabling or
disabling lingering when creating a given user account, otherwise
leaving the existing lingering configuration untouched.

tomeon committed:
nixos/users-groups: test user lingering management

tomeon committed:
nixos/users-groups: make `linger` trinary with `null` meaning "no action"

tomeon committed:
nixos/users-groups: mention user lingering management in release notes

tomeon committed:
maintainers: add tomeon

tomeon commented:
@ToxicFrog -- thank you for the feedback!

_Re_ (1) sounds good; thanks for the shot of confidence.
_Re_ (2)/(3), I've updated the `update-users-groups.pl` script to use an approach similar to that used by `switch-to-configuration.pl`; that is, [substitute occurrences of `@systemd@` with `${config.systemd.package}`](https://github.com/NixOS/nixpkgs/blob/55b7c77df29efdde3e93fe0930eb56d338e56934/nixos/modules/system/activation/activatable-system.nix#L36), and then look for systemd's utilities relative to that substituted path.  WDYT?

Yarny0 commented:
##### Reviewed points

- [x] changes are backward compatible *(compared to the situation before #260248)*
- [x] removed options are declared with `mkRemovedOptionModule` *(not applicable here)*
- [x] changes that are not backward compatible are documented in release notes *(not applicable here)*
- [x] module tests succeed *on x86-64*
- [x] options types are appropriate
- [x] options description is set
- [ ] options example is provided
- [x] documentation affected by the changes is updated *(none)*

##### Possible improvements

*   The mechanism does not work on bootup:

    > stage-2-init: disabling lingering for user(s) 'user2'
    > stage-2-init: System has not been booted with systemd as init system (PID 1). Can't operate.
    > stage-2-init: Failed to connect to bus: Host is down

    Instead of calling `loginctl`, one might add/remove the files in `/var/lib/systemd/linger` directly, but I'm wondering if changing lingering settings in the activation script is a good idea at all: Can we even be sure that `/var/lib/systemd/linger` is avaliable in the initrd?  It might get mounted later, by non-initrd systemd.

    Maybe it would be better to implement this as an ordinary systemd service:

    *   It would have to run before `logind`, but after `/var/lib/systemd/linger` gets mounted (can be ensured with `RequiresMountsFor=...`).
    *   It should be restarted on `nixos-rebuild switch`.
    *   It would need to be able to control lingering with `loginctl` (if `logind` is already running), or by adding/removing files in `/var/lib/systemd/linger` by hand (if `logind` is not running yet).
    *   As an ordinary system service, it would benefit from the parallelism of services at bootup.

*   The test succeeds, but it should probably be listed in `all-tests.nix`.  Given the problem at bootup, maybe we should also test a change of lingering settings at boot.


##### Comments

I strongly support this pull request: I have users that would like to control their lingering by themselves, and #260248 broke this.  Also I like the trinary design of the option: It's quick and clear!

tomeon committed:
nixos/users-groups: add fallback lingering management routines

that do not rely on systemd having booted or the systemd login manager
being up.  Detect whether the login manager is available by running
`loginctl show-sessions`; if this returns a nonzero status,  touch
`/var/lib/systemd/linger/${user}` when enabling lingering, and unlink
same when disabling lingering.

tomeon committed:
nixos/users: test boot-time lingering management

tomeon committed:
nixos/users: add diagnostics for disabling lingering of removed users

to distinguish between this and removing lingering for users who have
`{ linger = false; }`.

tomeon committed:
nixos/tests: harmonize user lingering approaches

tomeon commented:
@Yarny0 -- thanks for catching the boot-time issue.  I've added fallback lingering management methods to `update-users-groups.pl`, and added cases to the `user-lingering.nix` test to exercise these routines on initial boot and on reboot.  I've also added `user-lingering.nix` to `all-tests.nix`, so thanks for that as well.

_Re_ managing user lingering in an activation script: I implemented it like that in order to hook into the existing user creation and deletion logic.  I suppose it would be possible to make `update-users-groups.pl` write out a user lingering manifest to (say) `/var/lib/nixos/user-lingering`, along the lines of `/var/lib/nixos/{auto-subuid-map,declarative-groups,declarative-users,gid-map,uid-map}`.  Though presumably this approach is susceptible to concerns about the presence of the file path at boot time, as with `/var/lib/systemd/linger`.  Other options include removing features (e.g. only managing user lingering when `users.mutableUsers` is disabled).

aanderse commented:
purely informational but i'll mention https://github.com/NixOS/nixpkgs/pull/264879 as a landing point for you to look at - why? there are people who have intentions involving the scripts you're working on here

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/80

tomeon commented:
@ofborg test user-lingering
@ofborg test docker-rootless
@ofborg test jotta-cli
@ofborg test livebook-service
@ofborg test maestral
@ofborg test monado
@ofborg test systemd-user-tmpfiles-rules

Yarny0 commented:
It seems the goal has now already been accomplished by https://github.com/NixOS/nixpkgs/pull/460311 .  @tomeon: If I haven't missed anything, we should probably close the pull request at hand.
